[Draft, Splicing] Partly handle splice_init & splice_ack messages

optout21 opened the pull request:
Handle splice_init & splice_ack messages, without further performing splicing.
One next piece for splicing.
Can come before #3137 .

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10897,28 +11106,26 @@ where
-			"Splicing not supported".to_owned(),
-			msg.channel_id.clone())), counterparty_node_id);
+		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(self);
+		let _ = handle_error!(self, self.internal_splice_init(&counterparty_node_id, msg), counterparty_node_id);
 	}
 
 	#[cfg(splicing)]
 	fn handle_splice_ack(&self, counterparty_node_id: PublicKey, msg: &msgs::SpliceAck) {
-		let _: Result<(), _> = handle_error!(self, Err(MsgHandleErrInternal::send_err_msg_no_close(
-			"Splicing not supported (splice_ack)".to_owned(),
-			msg.channel_id.clone())), counterparty_node_id);
+		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(self);
+		let _ = handle_error!(self, self.internal_splice_ack(&counterparty_node_id, msg), counterparty_node_id);
 	}
```
Should we use `optionally_notify` in case of errors?

optout21 replied:
I'm not aware of this, need to dig into

optout21 replied:
Done

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3810,72 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+	pub fn get_splice_init(&self, our_funding_contribution_satoshis: i64, signer_provider: &SP,
+		funding_feerate_perkw: u32, locktime: u32,
+	) -> msgs::SpliceInit {
+		if !self.is_outbound() {
+			panic!("Tried to initiate a splice on an inbound channel!");
+		}
```
Why is this disallowed?

optout21 replied:
Good point, the splice initiator can be any of the channel peers (probably copy-paste from regular channel open)

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3810,72 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+		// channel value.
+		// Note that channel_keys_id is supposed NOT to change
+		let funding_pubkey = {
+			// TODO: Funding pubkey generation requires the post channel value, but that is not known yet,
+			// the acceptor contribution is missing. There is a need for a way to generate a new funding pubkey,
+			// not based on the channel value
```
Seems the funding pubkey derivation is not dependent on the amount in `InMemorySigner`. But it is passed to its constructor and used later when signing. @TheBlueMatt Seems will need to refactor our signing code to take `channel_value_satoshis` when signing instead. Do you see any issues with that?

optout21 replied:
I dropped creating new funding pubkey (as it was re-creating the same value anyways), therefore no need for signer generation.

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8903,6 +8986,132 @@ where
+					}
+
+					let post_channel_value = PendingSpliceInfoPre::compute_post_value(pre_channel_value, msg.funding_contribution_satoshis, our_funding_contribution);
+					if post_channel_value < 1000 {
+						return Err(MsgHandleErrInternal::send_err_msg_no_close(format!(
+							"Post-splicing channel value must be at least 1000 satoshis. It was {}", post_channel_value,
+						), msg.channel_id));
+					}
```
I'm not sure of the source of this, I add explicit checking against the reserve values.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3831,64 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+		// depends on the channel value, so we create here a new funding pubkey with the new
+		// channel value.
+		// Note that channel_keys_id is supposed NOT to change
+		let funding_pubkey = {
```
Is there is really need for funding_pubkey? Needs revisiting.

optout21 replied:
I dropped creating new funding pubkey (as it was re-creating the same value anyways), therefore no need for signer generation.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3831,52 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+	pub fn get_splice_init(&self, our_funding_contribution_satoshis: i64,
+		funding_feerate_perkw: u32, locktime: u32,
+	) -> msgs::SpliceInit {
+		// Reuse the existing funding pubkey, in spite of the channel value changing
+		// (though at this point we don't know the new value yet, due tue the optional counterparty contribution)
+		// Note that channel_keys_id is supposed NOT to change
+		let funding_pubkey = self.get_holder_pubkeys().funding_pubkey.clone();
+		msgs::SpliceInit {
+			channel_id: self.channel_id,
+			funding_contribution_satoshis: our_funding_contribution_satoshis,
+			funding_feerate_perkw,
+			locktime,
+			funding_pubkey,
+			require_confirmed_inputs: None,
+		}
```
Ah, the spec says `SHOULD` be different. https://github.com/lightning/bolts/pull/1160/files/aece0d8f42310d7e7d128905cf18eb0414c02459#r1631039188

@TheBlueMatt Are we fine with reusing the funding pubkey here?

optout21 replied:
We can handle creating new funding pubkey separately.

optout21 replied:
/see comment below/

optout21 replied:
There is no TODO in this place, but rather in the channel signer. The issue is that channel signer takes a channel value at construction time, so on the surface it looks like the signer should be re-created when the value changes. But this is not the case, as the value is not used in key generation, only when signing commitment transactions. Signer should be changed to be able to change the channel value (without regenerating all the keys; or alternatively it should not store the value at all). Also, the funding pubkey should be changed, this should also be possible with the Signer (change the funding key only).
I have raised an issue for the Signer for this -- #3542.
I update the comment to better reflect this.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4106,6 +4110,84 @@ where
+						return Err(APIError::APIMisuseError {
+							err: format!(
+								"Post-splicing channel value cannot be negative. It was {} - {}",
+								pre_channel_value, -our_funding_contribution_satoshis
```
This line can still cause an overflow for `i64::MIN`.

optout21 replied:
Good catch, errors within the generation of the error message are quite unpleasant.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8931,6 +9013,138 @@ where
+					// Sanity check: capacity cannot decrease below 0
+					if (pre_channel_value as i64).saturating_add(msg.funding_contribution_satoshis) < 0 {
+						return Err(MsgHandleErrInternal::send_err_msg_no_close(format!(
+							"Post-splicing channel value cannot be negative. It was {} - {}", pre_channel_value, -msg.funding_contribution_satoshis,
```
Likewise here.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1179,6 +1178,35 @@ impl UnfundedChannelContext {
+#[derive(Clone)]
+pub(crate) struct PendingSpliceInfoPre {
+	pub our_funding_contribution: i64,
+	// TODO(splicing): Enable below fields
+	// pub funding_feerate_perkw: u32,
+	// pub locktime: u32,
+	// /// The funding inputs we will be contributing to the splice.
+	// /// TODO(splice): will be changed to TransactionU16LenLimited
+	// pub our_funding_inputs: Vec<(TxIn, Transaction)>,
```
Let's remove any commented out code. We shouldn't merge any of that.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3771,6 +3831,52 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+	/// Get the splice_ack message that can be sent in response to splice initiation.
+	#[cfg(splicing)]
+	pub fn get_splice_ack(&mut self, our_funding_contribution_satoshis: i64) -> Result<msgs::SpliceAck, ChannelError> {
+		// TODO(splicing): checks
```
Are adding these checks blocked on #3137?

optout21 replied:
Added check for channel state, removed TODO

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3386,6 +3424,28 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
 
+	/// Check that a proposed channel value meets the channel reserve requirements or violates them (below reserve)
+	#[cfg(any(dual_funding, splicing))]
+	pub fn check_channel_value_meets_reserve_requirements(&self, proposed_channel_value: u64) -> Result<(), ChannelError> {
+		let holder_selected_channel_reserve_satoshis = get_v2_channel_reserve_satoshis(
+			proposed_channel_value, self.holder_dust_limit_satoshis);
+		if proposed_channel_value < holder_selected_channel_reserve_satoshis {
+			return Err(ChannelError::Warn(format!(
+				"Proposed channel value below reserve mandated by holder, {} vs {}",
+				proposed_channel_value, holder_selected_channel_reserve_satoshis,
+			)));
+		}
+		let counterparty_selected_channel_reserve_satoshis = get_v2_channel_reserve_satoshis(
+			proposed_channel_value, self.counterparty_dust_limit_satoshis);
+		if proposed_channel_value < counterparty_selected_channel_reserve_satoshis {
+			return Err(ChannelError::Warn(format!(
+				"Proposed channel value below reserve mandated by counterparty, {} vs {}",
+				proposed_channel_value, counterparty_selected_channel_reserve_satoshis,
+			)));
+		}
```
Not quite sure I understand these checks. Don't we want to make sure the balance (not the total value) doesn't violate the reserve requirements?

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1214,6 +1236,10 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
 
+	/// Info about an in-progress, pending splice (if any), on the pre-splice channel
+	#[cfg(splicing)]
+	pub(crate) pending_splice_pre: Option<PendingSpliceInfoPre>,
```
Is this only relevant to funded channels? Should we move it there instead?

optout21 replied:
You mean the ChannelPhase? I would put there only larger structs, not individual fields.
This field is also used to carry state from `splice_channel` till `channel_ack`, and during that time the channel is not actually splicing yet.

jkczyz replied:
Directly in `Channel` which is held by the `Funded` variant of `ChannelPhase`.

optout21 replied:
I have considered this, and changed it accordingly.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3386,6 +3418,29 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
 	}
 
+	/// Check that a balance value meets the channel reserve requirements or violates them (below reserve).
+	/// The channel value is an input, so that this can be used for checks with new planned channel value.
```
Not sure if the reader would know that "new planned channel value" pertains to splicing. Could you clarify the comment a bit?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8931,6 +9013,140 @@ where
+
+					let post_channel_value = PendingSpliceInfoPre::compute_post_value(pre_channel_value, msg.funding_contribution_satoshis, our_funding_contribution);
+
+					// Early check for reserve requirement, assuming maximum balance of full channel value
+					// This will also be checked later at tx_complete
+					let _res = chan.context.check_balance_meets_reserve_requirements(post_channel_value, post_channel_value)
+						.map_err(|e| MsgHandleErrInternal::from_chan_no_close(e, msg.channel_id))?;
```
Hmmm... why don't we use the actual channel balance given we have it? Since we are handling the `splice_init` here, we should know the amount we'll send in the `splice_ack`.

Or at least we know it is zero right now. Will the user eventually be given an event to ask if they want to contribute to the splice?

optout21 replied:
Good point. Changed to use computed new balance values.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8931,6 +9013,140 @@ where
+
+					// Check if a splice has been initiated already.
+					// Note: this could be handled more nicely, and support multiple outstanding splice's, the incoming splice_ack matters anyways.
+					if let Some(splice_info) = &chan.context.pending_splice_pre {
+						return Err(MsgHandleErrInternal::send_err_msg_no_close(format!(
+							"Channel has already a splice pending, channel id {}, contribution {}",
+							msg.channel_id, splice_info.our_funding_contribution,
+						), msg.channel_id));
+					}
```
@TheBlueMatt Could you remind us which checks should go in `channel` module vs `channelmanager`. I'm tempted to say we should keep `pending_splice_pre` private and only access it from `channel` but I'm unsure.

optout21 replied:
I reorganized to move all single-channel check and handling into `Channel` struct (`channel.rs`). The layout here followed the channel opening implementation, but it seems cleaner this way.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8931,6 +9013,140 @@ where
+			hash_map::Entry::Occupied(mut chan) => {
+				if let ChannelPhase::Funded(chan) = chan.get_mut() {
+					// check if splice is pending
+					let pending_splice = if let Some(pending_splice) = &chan.context.pending_splice_pre {
+						// Note: this is incomplete (their funding contribution is not set)
+						pending_splice.clone()
+					} else {
+						return Err(MsgHandleErrInternal::send_err_msg_no_close("Channel is not in pending splice".to_owned(), msg.channel_id));
+					};
+
+					let pre_channel_value = chan.context.get_value_satoshis();
+					let post_channel_value = PendingSpliceInfoPre::compute_post_value(pre_channel_value, pending_splice.our_funding_contribution, msg.funding_contribution_satoshis);
+
+					// Early check for reserve requirement, assuming maximum balance of full channel value
+					// This will also be checked later at tx_complete
+					let _res = chan.context.check_balance_meets_reserve_requirements(post_channel_value, post_channel_value)
+						.map_err(|e| MsgHandleErrInternal::from_chan_no_close(e, msg.channel_id))?;
```
Similarly here? Seems there's a lot of back-and-forth with `channel` when it can be done in there instead.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8931,6 +9013,140 @@ where
+						), msg.channel_id));
+					}
+
+					// TODO(splicing): Below step should come after channel phase change (after re-adding to channel map)
```
Do we ever remove it from the channel map? Was thinking we were just keeping it in the funded phase.

optout21 replied:
Not as of now, removed TODO comment

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8931,6 +9013,140 @@ where
+					// check if splice is pending
+					let pending_splice = if let Some(pending_splice) = &chan.context.pending_splice_pre {
+						// Note: this is incomplete (their funding contribution is not set)
+						pending_splice.clone()
```
Why does this need to be cloned?

optout21 replied:
It doesn't, my bad

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1179,6 +1178,29 @@ impl UnfundedChannelContext {
+/// Info about a pending splice, used in the pre-splice channel
+#[cfg(splicing)]
+#[derive(Clone)]
+pub(crate) struct PendingSpliceInfoPre {
```
True. Now it can be even private, see below.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1179,6 +1178,29 @@ impl UnfundedChannelContext {
+
+#[cfg(splicing)]
+impl PendingSpliceInfoPre {
+	fn add_checked(base: u64, delta: i64) -> u64 {
+		if delta >= 0 {
+			base.saturating_add(delta as u64)
+		} else {
+			base.saturating_sub(delta.abs() as u64)
+		}
+	}
```
Inlining would mean duplicating a long expression 3x (the delta), so I only marked this method #[inline]

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -7350,6 +7439,24 @@ impl<SP: Deref> Channel<SP> where
+		-> Result<msgs::SpliceAck, ChannelError>
+		where ES::Target: EntropySource, L::Target: Logger
+	{
+		if !self.context.is_outbound() {
```
Indeed, done

optout21 commented:
I have reorganized code, moved most single-channel checks and handling logic into `Channel` struct from `ChannelManager`.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -3615,6 +3648,33 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+	/// Check that a balance value meets the channel reserve requirements or violates them (below reserve).
+	/// The channel value is an input as opposed to using from self, so that this can be used in case of splicing
+	/// to checks with new channel value (before being comitted to it).
+	#[cfg(any(dual_funding, splicing))]
```
CI is sad because `dual_funding` has been removed.

optout21 commented:
Rebased (post #3137) and squashed

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4118,6 +4118,65 @@ where
+	#[cfg(splicing)]
+	pub fn splice_channel(
+		&self, channel_id: &ChannelId, counterparty_node_id: &PublicKey, our_funding_contribution_satoshis: i64,
+		_our_funding_inputs: Vec<(TxIn, Transaction)>, funding_feerate_perkw: u32, locktime: u32,
```
Checking is done later when interactive negotiation is started, but early check is preferred, so I've added this check.
Storing the inputs will come in the next step.

optout21 replied:
Will come in next step. `InteractiveTxConstructor` is still missing in dual funding.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1182,6 +1181,30 @@ impl UnfundedChannelContext {
+/// Info about a pending splice, used in the pre-splice channel
+#[cfg(splicing)]
+#[derive(Clone)]
+struct PendingSpliceInfoPre {
```
Trying to think of a better name for this. Is it necessary to divide this into "pre" and "post" structs? Thinking we can simply call this `PendingSplice`.

optout21 replied:
Renamed

optout21 commented:
Rebased and squashed

optout21 commented:
Rebased (minor merge conflict)

optout21 commented:
Rebased (post Inbound/OutboundV2Channel merge #3498 )

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -7781,6 +7866,135 @@ impl<SP: Deref> Channel<SP> where
+		our_funding_inputs: Vec<(TxIn, Transaction)>, funding_feerate_perkw: u32, locktime: u32,
+	) -> Result<msgs::SpliceInit, ChannelError> {
+		// Check if a splice has been initiated already.
+		// Note: this could be handled more nicely, and support multiple outstanding splice's, the incoming splice_ack matters anyways.
```
Clarified the comment

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11326,28 +11471,42 @@ where
-			msg.channel_id.clone())), counterparty_node_id);
+		let _persistence_guard = PersistenceNotifierGuard::optionally_notify(self, || {
+			let res = self.internal_splice_init(&counterparty_node_id, msg);
+			let persist = match &res {
+				Err(e) if e.closes_channel() => NotifyOption::DoPersist,
+				Err(_) => NotifyOption::SkipPersistHandleEvents,
+				Ok(()) => NotifyOption::SkipPersistNoEvents,
+			};
```
TODO was missing, added

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -3584,6 +3607,33 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+	/// The channel value is an input as opposed to using from self, so that this can be used in case of splicing
+	/// to checks with new channel value (before being comitted to it).
+	#[cfg(splicing)]
+	pub fn check_balance_meets_reserve_requirements(&self, balance: u64, channel_value: u64) -> Result<(), ChannelError> {
```
OK, I was not aware that this reserve condition is v2 specific.

optout21 commented:
> Can you add a TODO and an issue to track it to make sure we don't forget? I was just about to comment the same thing until I saw this resolved comment.

```
// Reuse the existing funding pubkey, in spite of the channel value changing
// (though at this point we don't know the new value yet, due tue the optional counterparty contribution)
```

Regarding this (for some reason couldn't reply to this comment in-place):
There is no TODO in this place, but rather in the channel signer. The issue is that channel signer takes a channel value at construction time, so on the surface it looks like the signer should be re-created when the value changes. But this is not the case, as the value is not used in key generation, only when signing commitment transactions. Signer should be changed to be able to change the channel value (without regenerating all the keys; or alternatively it should not store the value at all). Also, the funding pubkey should be changed, this should also be possible with the Signer (change the funding key only).
I have raised an issue for the Signer for this -- #3542.
I update the comment to better reflect this.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -7781,6 +7866,135 @@ impl<SP: Deref> Channel<SP> where
+		let sum_input: i64 = our_funding_inputs.into_iter().map(
+			|(txin, tx)| tx.output.get(txin.previous_output.vout as usize).map(|tx| tx.value.to_sat() as i64).unwrap_or(0)
+		).sum();
+		if sum_input < our_funding_contribution_satoshis {
```
This is a pre-check, additional check should be done later (when all contributions and outputs are known). Fees could be estimated here, but I think it's ok to leave the check for later.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -7781,6 +7866,135 @@ impl<SP: Deref> Channel<SP> where
+		// TODO(splicing): Store msg.funding_pubkey
+		// TODO(splicing): Apply start of splice (splice_start)
+
+		let splice_ack_msg = self.context.get_splice_ack(our_funding_contribution_satoshis);
```
I am a bit confused about this one. Do you suggest some addition to the code?

optout21 commented:
I close the remaining two outstanding review comments. Note that this now depends/includes #3558 (I'll take care if merge is needed).

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8028,6 +8130,156 @@ impl<SP: Deref> FundedChannel<SP> where
+		// Input witness weight, extended with weight for spending old funding
+		let total_witness_weight = Weight::from_wu(
+			witness_weight.to_wu()
+				.saturating_add(2 * P2WPKH_WITNESS_WEIGHT)
```
Yes, it's for the 2 signature witnesses. Is it incorrect?
I will add a comment about it.

optout21 replied:
Indeed, I was sloppy here! (4 bytes were omitted, 218).  I've added new consts to have the correct value.

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4206,6 +4206,65 @@ where
+	#[cfg(splicing)]
+	pub fn splice_channel(
+		&self, channel_id: &ChannelId, counterparty_node_id: &PublicKey, our_funding_contribution_satoshis: i64,
+		our_funding_inputs: Vec<(TxIn, Transaction)>, witness_weight: Weight,
```
Good idea, logical, but it's already used like this in dual funding (👀 @dunxen ). Not sure how to handle it:
- use the existing style for splicing as well, handle separately to change for both dual f. and splicing
- do it like that in this PR for splicing, handle for dual funding separately -- may need to convert if same helpers are to be reused
- change in this PR for dual funding as well
I prefer not to change non-strictly-splicing related parts in this PR.

optout21 replied:
Did the change to `(TxIn, Transaction, Weight)` for the splicing part (added TODO to dual funding part).

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8028,6 +8130,156 @@ impl<SP: Deref> FundedChannel<SP> where
+
+		let pre_channel_value = self.context.get_value_satoshis();
+		// Sanity check: capacity cannot decrease below 0
+		if (pre_channel_value as i64).saturating_add(our_funding_contribution_satoshis) < 0 {
```
True, replaced with a TODO (for the case when splice out is done)

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8028,6 +8130,156 @@ impl<SP: Deref> FundedChannel<SP> where
+		// Pre-check that inputs are sufficient to cover our contribution.
+		// Note: fees are not taken into account here.
+		let sum_input: u64 = our_funding_inputs.iter().map(
+			|(txin, tx)| tx.output.get(txin.previous_output.vout as usize).map(|tx| tx.value.to_sat()).unwrap_or(0)
```
This is outdated now; checked in only one place.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8028,6 +8130,156 @@ impl<SP: Deref> FundedChannel<SP> where
+		// Sanity check: capacity cannot decrease below 0
+		if (pre_channel_value as i64)
+			.saturating_add(their_funding_contribution_satoshis)
+			.saturating_add(our_funding_contribution_satoshis) < 0
```
Same (removed, replaced with TODO to check once acceptor contribution is added)

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -4290,6 +4343,38 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+
+	/// Get the splice message that can be sent during splice initiation.
+	#[cfg(splicing)]
+	pub fn get_splice_init(&self, our_funding_contribution_satoshis: i64,
```
The logic was that if it uses only fields from `ChannelContext`, it is defined there, but logically it makes sense only on `FundedChannel`. I have moved it -- this way connected methods are closer in the source file, plus it means less impact when/if `ChannelContext` is broken up.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8028,6 +8130,156 @@ impl<SP: Deref> FundedChannel<SP> where
+		let post_balance = PendingSplice::add_checked(self.context.value_to_self_msat, our_funding_contribution);
+		// Pre-check for reserve requirement, assuming maximum balance of full channel value
+		// This will also be checked later at tx_complete
+		let _res = self.context.check_balance_meets_v2_reserve_requirements(post_balance, post_channel_value)?;
```
Check adapted.

optout21 commented:
Please kindly re-consider.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8114,6 +8236,157 @@ impl<SP: Deref> FundedChannel<SP> where
+
+		// Check that inputs are sufficient to cover our contribution.
+		// Extra common weight is the weight for spending the old funding
+		let extra_input_weight = Weight::from_wu(FUNDING_TRANSACTION_WITNESS_WEIGHT);
```
In that case this method cannot be used by V2 channel open. The old funding tx input is only present in the splicing case.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -4486,30 +4571,64 @@ pub(super) fn calculate_our_funding_satoshis(
+		}
+	}
+
+	// If the inputs are enough to cover intended contribution amount, with fees even when
+	// there is a change output, we are fine.
+	// If the inputs are less, but enough to cover intended contribution amount, with
+	// (lower) fees with no change, we are also fine (change will not be generated).
+	// So it's enough to check considering the lower, no-change fees.
+	// Note: dust limit is not relevant in this check.
+
+	let minimal_input_amount_needed = contribution_amount.saturating_add(estimated_fee as i64);
```
My thinking here is that in obvious user error don't bother to make calls to the counterparty, but fail as early as possible, as that's better. Don't you think?

optout21 replied:
I see, but that's not really possible, for two reasons:
- as of this PR (+current main), dual funding negotiation is not yet implemented.
- with dual funding negotiation implemented, the decision to create a change output is done later, at `splice_ack` time, just before setting up dual funding negotiation, in `begin_interactive_funding_tx_construction`, and that cannot be called twice. Final fee is known at `tx_completion` time.

So to be able to do this, these would be needed:

- add a way to check whether change output is needed, without a change in state.
- implement dual funding negotiation preparation.

I'm not changing this in this PR.

optout21 replied:
Add'l info: a `calculate_change_output_value` method is being planned in #3443   https://github.com/lightningdevkit/rust-lightning/pull/3443/files#diff-c92e994fd9949ebcb6a8d9ec76553e8497ee223b707b51905e35017b80124c77R1671

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4231,6 +4231,68 @@ where
+		our_funding_inputs: Vec<(TxIn, Transaction, Weight)>,
+		funding_feerate_per_kw: u32, locktime: Option<u32>,
+	) -> Result<(), APIError> {
+		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(self);
```
Changes to optionally_notify(), but a review here would be helpful.

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4231,6 +4231,68 @@ where
+					let msg = chan.splice_channel(our_funding_contribution_satoshis, our_funding_inputs, funding_feerate_per_kw, locktime)
+						.map_err(|err| APIError::APIMisuseError {
+							err: format!(
+								"Cannot initiate Splicing, {}, channel ID {}", err, channel_id
```
This error wrapping is no longer present, but updated error msgs one level upsteam.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8114,6 +8236,157 @@ impl<SP: Deref> FundedChannel<SP> where
+			)));
+		}
+
+		if !matches!(self.context.channel_state, ChannelState::ChannelReady(_)) {
```
Done

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8114,6 +8236,157 @@ impl<SP: Deref> FundedChannel<SP> where
+			)));
+		}
+
+		if !matches!(self.context.channel_state, ChannelState::ChannelReady(_)) {
```
Done

optout21 commented on lightning/src/ln/functional_tests_splice.rs:
```diff
@@ -0,0 +1,331 @@
+	// Have initiator_node initiate a channel to acceptor_node with aforementioned parameters
+	let channel_id_temp1 = initiator_node
+		.node
+		.create_channel(
```
This indeed simplified the V1 channel setup a lot. (Bg: I was mostly testing splicing with V2 channels, and there we didn't have ready utility methods).

optout21 commented:
Addressed review comments, except one (try_channel_entry ...)

optout21 commented:
Addressed review comments, please reconsider.

optout21 commented:
Note: new conflict, has to be merged properly with #3558 , will do.

optout21 commented:
> Note: new conflict, has to be merged properly with #3558 , will do.

Done

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4256,6 +4259,81 @@ where
+				Ok(p) => p,
+				Err(e) => {
+					res = Err(e);
+					return NotifyOption::SkipPersistNoEvents;
```
Good idea, done

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11484,28 +11650,42 @@ where
+			let persist = match &res {
+				Err(e) if e.closes_channel() => NotifyOption::DoPersist,
+				Err(_) => NotifyOption::SkipPersistHandleEvents,
+				Ok(()) => NotifyOption::SkipPersistNoEvents,
```
Thx, done

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4256,6 +4259,81 @@ where
+	#[cfg(splicing)]
+	pub fn splice_channel(
+		&self, channel_id: &ChannelId, counterparty_node_id: &PublicKey, our_funding_contribution_satoshis: i64,
+		our_funding_inputs: Vec<(TxIn, Transaction, Weight)>,
```
It would be possible, but what would be the benefit?
In v1 funding the caller must create the full transaction, and that's not possible to do at the beginning of the process, hence the solution with the callback. But that's not the case for splicing, and available inputs can be collected upfront.

optout21 replied:
Probably, but this is further up in the road. Dual funding channel creation is not yet implemented, and the closest version where it is implemented, inputs are supplied directly in the `create...()` call by the library user.
Can be seen here:  https://github.com/lightningdevkit/rust-lightning/pull/2302/files#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR3018  (not should directly go to line, if not, open `channelmanager.rs` line 3018)
I agree that this is up for discussion, but not in this PR.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -3928,6 +3955,34 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+	/// to checks with new channel value (before being comitted to it).
+	#[cfg(splicing)]
+	pub fn check_balance_meets_v2_reserve_requirements(&self, self_balance: u64, counterparty_balance: u64, channel_value: u64) -> Result<(), ChannelError> {
+		if self_balance > 0 {
```
I don't understand, we don't have control over the pre-splice value, why to check if pre-splice value was above reserve? If it was below, we should allow to be below post-splice as well?

optout21 replied:
OK, implemented it that way.

optout21 commented:
Rebased following FundingScope (#3592)
New comments addressed, fixed.
No additional change planned on my side.
Awaiting clarifying/closing remaining pending comments.

optout21 commented:
Yet another rebase, this time #3588 (quiescence) -- only superficial merge, functionality not integrated (prefer handling that in a fresh PR)

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4278,6 +4281,88 @@ where
+	) -> Result<(), APIError> {
+		let mut res = Ok(());
+		PersistenceNotifierGuard::optionally_notify(self, || {
+			let (result, notify_option) = self.internal_splice_channel(
```
Nice, done

jkczyz commented:
Is there any chance you could rearrange the commits such the fixups are adjacent to the commits that they are fixing (and named to imply as much)? Otherwise, it is difficult to review.

optout21 commented:
> Is there any chance you could rearrange the commits such the fixups are adjacent to the commits that they are fixing (and named to imply as much)? Otherwise, it is difficult to review.

There is on change commit and 26 various changes in response to reviews. I'm not sure how could meaningfully group them. I'll try to think of something. In the current state, probably best is to review the cummulative change.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4278,6 +4281,91 @@ where
+			hash_map::Entry::Occupied(mut chan_phase_entry) => {
+				let locktime = locktime.unwrap_or(self.current_best_block().height);
+				if let Some(chan) = chan_phase_entry.get_mut().as_funded_mut() {
+					let msg = match chan.splice_channel(our_funding_contribution_satoshis, our_funding_inputs, funding_feerate_per_kw, locktime) {
```
Use `?` notation for returning.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4278,6 +4281,91 @@ where
+			let result = self.internal_splice_channel(
+				channel_id, counterparty_node_id, our_funding_contribution_satoshis, &our_funding_inputs, funding_feerate_per_kw, locktime
+			);
+			res = result;
```
Is this needed?

optout21 commented:
Combined previous commits to decrease commit bloat

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -4068,6 +4095,68 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
+		} else {
+			((balance >= channel_reserve), channel_reserve)
+		}
+	}
```
You mean like this:

```
	pub fn check_splice_balance_meets_v2_reserve_requirement_noerr(pre_balance: u64, post_balance: u64, pre_channel_value: u64, post_channel_value: u64, dust_limit: u64) -> Result<(), u64> {
		if post_balance == 0 {
			// 0 balance is fine
			Ok(())
		} else {
			let post_channel_reserve = get_v2_channel_reserve_satoshis(post_channel_value, dust_limit);
			if post_balance >= post_channel_reserve {
				Ok(())
			} else {
				// post is not OK, check pre
				if pre_balance == 0 {
					// pre OK, post not -> not
					Err(post_channel_reserve)
				} else {
					let pre_channel_reserve = get_v2_channel_reserve_satoshis(pre_channel_value, dust_limit);
					if pre_balance >= pre_channel_reserve {
						// pre OK, post not -> not
						Err(post_channel_reserve)
					} else {
						// post not OK, but so was pre -> OK
						Ok(())
					}
				}
			}
		}
	}
```
?

It has a bit more repetition, less symmetry, and more nesting. Nesting is only 4 levels, which is still OK.
I consider the broken-up version cleaner, but I can change, np.

jkczyz replied:
> It has a bit more repetition, less symmetry, and more nesting. Nesting is only 4 levels, which is still OK.
I consider the broken-up version cleaner, but I can change, np.

Yeah, though nesting can reduced with early returns. I'd defer to @wpaulino on what is actually needed, though.

optout21 commented:
Pushed few changes in response to recent review comments.

There is one outstanding comment at the moment, reg. computing the counterparty balance, from @wpaulino  https://github.com/lightningdevkit/rust-lightning/pull/3407#discussion_r1970725164
-- I'll look into refining that soon.

optout21 commented:
Removed the channel reserve requirements checks; to be handled in follow-up #3641 .

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8277,6 +8421,162 @@ impl<SP: Deref> FundedChannel<SP> where
+		let post_balance_counterparty = post_channel_value.saturating_sub(post_balance_self);
+		// Pre-check for reserve requirement
+		// This will also be checked later at tx_complete
+		let _res = self.context.check_splice_balances_meet_v2_reserve_requirements(pre_balance_self, post_balance_self, pre_balance_counterparty, post_balance_counterparty, pre_channel_value, post_channel_value)?;
```
--> #3641

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -4667,6 +4694,58 @@ fn estimate_v2_funding_transaction_fee(
+		// consider the weight of the witness needed for spending the old funding transaction
+		total_input_witness_weight += Weight::from_wu(FUNDING_TRANSACTION_WITNESS_WEIGHT);
+	}
+	let estimated_fee = estimate_v2_funding_transaction_fee(is_initiator, funding_inputs.len(), total_input_witness_weight, funding_feerate_sat_per_1000_weight);
```
Good catch! In the if just above, the witness weight was added for the old funding input (for the splice initiator), but the weight of the input itself was not added.

optout21 commented:
Review comments from @wpaulino addressed.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -1719,6 +1722,30 @@ impl FundingScope {
+		}
+	}
+
+	/// Compute the post-splice channel value from the pre-splice values and the peer contributions
+	pub fn compute_post_value(pre_channel_value: u64, our_funding_contribution: i64, their_funding_contribution: i64) -> u64 {
+		Self::add_checked(pre_channel_value, our_funding_contribution.saturating_add(their_funding_contribution))
+	}
```
Looks like this isn't needed in this PR anymore. Same with `add_checked`.

optout21 replied:
Indeed. Interesting, `cargo t -p lightning <pattern>` did not show these unused warnings, while `cargo t -p lightning` does (not clear to me why).
These are removed now.

jkczyz replied:
You'll need to use `RUSTFLAGS="--cfg=splicing" cargo +1.63.0 test -p lightning`

optout21 replied:
I've been using `splicing` flag, but interestingly the difference is between:
```
RUSTFLAGS="--cfg=splicing" cargo t -p lightning splic
```
and
```
RUSTFLAGS="--cfg=splicing" cargo t -p lightning
```
The former did not show the unused warning, the latter did. Oh, well. That's why we have CI's :)

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -8323,6 +8407,149 @@ impl<SP: Deref> FundedChannel<SP> where
+		// (Cannot test for miminum required post-splice channel value)
+
+		// Check that inputs are sufficient to cover our contribution.
+		// Extra common weight is the weight for spending the old funding
```
Outdated comment.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -12684,4 +12921,177 @@ mod tests {
+
+	#[cfg(splicing)]
+	#[test]
+	fn test_check_v2_funding_inputs_sufficient() {
```
Not updated with last commit, my bad.

jkczyz reviewed (commented):
LGTM. @wpaulino Good to squash the fixups?

optout21 commented:
Squashed into a single commit, prepared for merging.

jkczyz reviewed (approved):
Still waiting on CI to start... :(

optout21 commented:
Somehow CI got mixed up, showing failures for **earlier** commit...  I'm not sure how to re-trigger.

![image](https://github.com/user-attachments/assets/585ab88e-d85b-4746-93d1-d93410507ad9)

![image](https://github.com/user-attachments/assets/5d77f391-4f32-4cae-bcf8-63639f5a20f9)

Commit ID is different. I know the issue being reported (conflict caused by #3016) but fixed previously.

optout21 commented:
Did a commit amend & force push, let's see CI's response...

optout21 commented:
For some reason (which I don't understand), when in the CI this PR branch is merged onto current main, a bunch of previous rebase merges get reverted, and compilation fails. Even though the PR branch was rebased to recent and has no conflicts. I have recreated the commit completely anew, hopefully this helps.

jkczyz commented:
> For some reason (which I don't understand), when in the CI this PR branch is merged onto current main, a bunch of previous rebase merges get reverted, and compilation fails. Even though the PR branch was rebased to recent and has no conflicts. I have recreated the commit completely anew, hopefully this helps.

Github shows that your branch is 33 commits behind. If I rebase your PR on upstream `main`, I can reproduce the CI failure with `RUSTFLAGS="--cfg=splicing" cargo +1.63.0 check -p lightning`. Seems you need to account for #3594.

jkczyz commented:
> Github shows that your branch is 33 commits behind. If I rebase your PR on upstream `main`, I can reproduce the CI failure with `RUSTFLAGS="--cfg=splicing" cargo +1.63.0 check -p lightning`. Seems you need to account for #3594.

Rebased and fixed the compilation error locally. CI running now.
